Add statusUpdate code

cammellos opened the pull request:
This commit adds a new code `statusUpdate = 2` which replaces:

`powRequirementCode`
`bloomFilterExCode`
`rateLimitingCode`

Tested against status-react and currently deployed on `eth.test` fleet

Closes https://github.com/status-im/status-go/issues/1853

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (81)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 1375b656 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/1/) | 2020-02-17 15:34:33 | ~1 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200217-153328-1375b6.zip) |
| :heavy_check_mark: | 1375b656 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/1/) | 2020-02-17 15:39:30 | ~6 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200217-153327-1375b6.zip) |
| :heavy_check_mark: | 1375b656 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/1/) | 2020-02-17 15:42:16 | ~8 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200217-153325-1375b6.aar) |
| | | | | | | |
| :heavy_check_mark: | 3c678bb9 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/2/) | 2020-02-17 15:35:15 | ~39 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200217-153435-3c678b.zip) |
| :heavy_check_mark: | 3c678bb9 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/2/) | 2020-02-17 15:42:32 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200217-153932-3c678b.zip) |
| :heavy_check_mark: | 3c678bb9 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/2/) | 2020-02-17 15:48:15 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200217-154218-3c678b.aar) |
| | | | | | | |
| :heavy_check_mark: | 16b62704 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/3/) | 2020-02-17 17:14:05 | ~47 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200217-171318-16b627.zip) |
| :heavy_check_mark: | 16b62704 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/3/) | 2020-02-17 17:15:50 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200217-171317-16b627.zip) |
| :heavy_check_mark: | 16b62704 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/3/) | 2020-02-17 17:20:30 | ~7 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200217-171315-16b627.aar) |
| | | | | | | |
| :heavy_check_mark: | 36f12158 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/4/) | 2020-02-17 17:19:59 | ~31 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200217-171928-36f121.zip) |
| :heavy_check_mark: | 36f12158 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/4/) | 2020-02-17 17:21:48 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200217-171926-36f121.zip) |
| :heavy_check_mark: | 36f12158 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/4/) | 2020-02-17 17:28:28 | ~7 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200217-172032-36f121.aar) |
| | | | | | | |
| :heavy_check_mark: | a8c54932 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/5/) | 2020-02-18 09:13:15 | ~51 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200218-091223-a8c549.zip) |
| :heavy_check_mark: | a8c54932 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/5/) | 2020-02-18 09:15:09 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200218-091222-a8c549.zip) |
| :heavy_check_mark: | a8c54932 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/5/) | 2020-02-18 09:20:00 | ~7 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200218-091221-a8c549.aar) |
| | | | | | | |
| :heavy_check_mark: | 5ef6d031 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/6/) | 2020-02-18 09:46:44 | ~34 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200218-094609-5ef6d0.zip) |
| :heavy_check_mark: | 5ef6d031 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/6/) | 2020-02-18 09:48:47 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200218-094607-5ef6d0.zip) |
| :heavy_check_mark: | 5ef6d031 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/6/) | 2020-02-18 09:52:28 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200218-094606-5ef6d0.aar) |
| | | | | | | |
| :heavy_check_mark: | 2aadff74 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/7/) | 2020-02-18 10:14:15 | ~1 min | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200218-101253-2aadff.zip) |
| :heavy_check_mark: | 2aadff74 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/7/) | 2020-02-18 10:15:46 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200218-101251-2aadff.zip) |
| :heavy_check_mark: | 2aadff74 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/7/) | 2020-02-18 10:20:54 | ~8 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200218-101248-2aadff.aar) |
| | | | | | | |
| :heavy_check_mark: | 59edc786 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/8/) | 2020-02-18 12:27:10 | ~46 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200218-122623-59edc7.zip) |
| :heavy_check_mark: | 59edc786 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/8/) | 2020-02-18 12:29:01 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200218-122621-59edc7.zip) |
| :heavy_check_mark: | 59edc786 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/8/) | 2020-02-18 12:32:50 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200218-122618-59edc7.aar) |
| | | | | | | |
| :x: | 99285159 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/9/) | 2020-02-18 13:00:59 | ~16 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/9/consoleText) |
| :x: | 99285159 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/9/) | 2020-02-18 13:01:10 | ~25 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/9/consoleText) |
| :heavy_check_mark: | 99285159 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/9/) | 2020-02-18 13:01:43 | ~55 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200218-130047-992851.zip) |
| | | | | | | |
| :x: | 8a176a04 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/10/) | 2020-02-18 17:00:17 | ~17 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/10/consoleText) |
| :x: | 8a176a04 | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/10/) | 2020-02-18 17:00:30 | ~26 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/10/consoleText) |
| :heavy_check_mark: | 8a176a04 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/10/) | 2020-02-18 17:00:47 | ~38 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200218-170007-8a176a.zip) |
| | | | | | | |
| :x: | 8422b029 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/11/) | 2020-02-19 10:00:48 | ~18 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/11/consoleText) |
| :x: | 8422b029 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/11/) | 2020-02-19 10:01:02 | ~29 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/11/consoleText) |
| :heavy_check_mark: | 8422b029 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/11/) | 2020-02-19 10:01:12 | ~36 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200219-100035-8422b0.zip) |
| | | | | | | |
| :heavy_check_mark: | fcdf9056 | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/12/) | 2020-02-19 10:25:07 | ~41 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200219-102425-fcdf90.zip) |
| :heavy_check_mark: | fcdf9056 | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/12/) | 2020-02-19 10:27:05 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200219-102422-fcdf90.zip) |
| :heavy_check_mark: | fcdf9056 | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/12/) | 2020-02-19 10:31:11 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200219-102418-fcdf90.aar) |
| | | | | | | |
| :heavy_check_mark: | 4ee76b63 | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/13/) | 2020-02-19 10:40:41 | ~37 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200219-104004-4ee76b.zip) |
| :heavy_check_mark: | 4ee76b63 | [#13](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/13/) | 2020-02-19 10:42:29 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200219-104000-4ee76b.zip) |
| :heavy_check_mark: | 4ee76b63 | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/13/) | 2020-02-19 10:46:09 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200219-103957-4ee76b.aar) |
| | | | | | | |
| :heavy_check_mark: | a9301c0a | [#14](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/14/) | 2020-02-19 11:24:16 | ~40 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200219-112336-a9301c.zip) |
| :heavy_check_mark: | a9301c0a | [#14](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/14/) | 2020-02-19 11:26:02 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200219-112333-a9301c.zip) |
| :heavy_check_mark: | a9301c0a | [#14](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/14/) | 2020-02-19 11:29:54 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200219-112331-a9301c.aar) |
| | | | | | | |
| :heavy_check_mark: | 654dfddf | [#15](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/15/) | 2020-02-19 11:30:47 | ~34 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200219-113013-654dfd.zip) |
| :heavy_check_mark: | 654dfddf | [#15](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/15/) | 2020-02-19 11:32:25 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200219-113009-654dfd.zip) |
| :heavy_check_mark: | 654dfddf | [#15](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/15/) | 2020-02-19 11:37:00 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200219-113007-654dfd.aar) |
| | | | | | | |
| :x: | d883010d | [#16](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/16/) | 2020-02-20 08:43:17 | ~23 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/16/consoleText) |
| :heavy_check_mark: | d883010d | [#16](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/16/) | 2020-02-20 08:43:37 | ~40 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200220-084256-d88301.zip) |
| :heavy_check_mark: | d883010d | [#16](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/16/) | 2020-02-20 08:45:43 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200220-084255-d88301.zip) |
| | | | | | | |
| :x: | 87643f3c | [#17](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/17/) | 2020-02-20 08:44:42 | ~14 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/17/consoleText) |
| :heavy_check_mark: | 87643f3c | [#17](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/17/) | 2020-02-20 08:45:00 | ~28 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200220-084431-87643f.zip) |
| :heavy_check_mark: | 87643f3c | [#17](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/17/) | 2020-02-20 08:48:23 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200220-084545-87643f.zip) |
| | | | | | | |
| :heavy_check_mark: | 4cb4fc37 | [#18](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/18/) | 2020-02-20 10:07:51 | ~36 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200220-100714-4cb4fc.zip) |
| :heavy_check_mark: | 4cb4fc37 | [#18](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/18/) | 2020-02-20 10:09:41 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200220-100711-4cb4fc.zip) |
| :heavy_check_mark: | 4cb4fc37 | [#18](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/18/) | 2020-02-20 10:13:41 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200220-100709-4cb4fc.aar) |
| | | | | | | |
| :heavy_check_mark: | 02c54230 | [#19](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/19/) | 2020-02-20 10:53:25 | ~39 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200220-105245-02c542.zip) |
| :heavy_check_mark: | 02c54230 | [#19](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/19/) | 2020-02-20 10:55:15 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200220-105243-02c542.zip) |
| :heavy_check_mark: | 02c54230 | [#19](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/19/) | 2020-02-20 10:59:16 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200220-105241-02c542.aar) |
| | | | | | | |
| :heavy_check_mark: | fd5f9849 | [#20](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/20/) | 2020-02-20 11:53:28 | ~37 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200220-115251-fd5f98.zip) |
| :heavy_check_mark: | fd5f9849 | [#20](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/20/) | 2020-02-20 11:55:21 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200220-115249-fd5f98.zip) |
| :heavy_check_mark: | fd5f9849 | [#20](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/20/) | 2020-02-20 11:59:02 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200220-115247-fd5f98.aar) |
| | | | | | | |
| :heavy_check_mark: | a0e3becd | [#21](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/21/) | 2020-02-21 11:32:59 | ~53 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200221-113205-a0e3be.zip) |
| :heavy_check_mark: | a0e3becd | [#21](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/21/) | 2020-02-21 11:34:31 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200221-113203-a0e3be.zip) |
| :heavy_check_mark: | a0e3becd | [#21](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/21/) | 2020-02-21 11:38:28 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200221-113200-a0e3be.aar) |
| | | | | | | |
| :heavy_check_mark: | a186f53b | [#22](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/22/) | 2020-02-21 11:33:39 | ~27 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200221-113311-a186f5.zip) |
| :heavy_check_mark: | a186f53b | [#22](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/22/) | 2020-02-21 11:36:56 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200221-113433-a186f5.zip) |
| :heavy_check_mark: | a186f53b | [#22](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/22/) | 2020-02-21 11:44:35 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200221-113830-a186f5.aar) |
| | | | | | | |
| :heavy_check_mark: | b139290c | [#23](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/23/) | 2020-02-21 12:09:36 | ~39 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200221-120856-b13929.zip) |
| :heavy_check_mark: | b139290c | [#23](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/23/) | 2020-02-21 12:11:42 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200221-120854-b13929.zip) |
| :heavy_check_mark: | b139290c | [#23](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/23/) | 2020-02-21 12:15:04 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200221-120852-b13929.aar) |
| | | | | | | |
| :x: | c96d695a | [#24](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/24/) | 2020-02-21 12:17:07 | ~14 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/24/consoleText) |
| :heavy_check_mark: | c96d695a | [#24](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/24/) | 2020-02-21 12:17:28 | ~30 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200221-121656-c96d69.zip) |
| :heavy_check_mark: | c96d695a | [#24](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/24/) | 2020-02-21 12:19:42 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200221-121655-c96d69.zip) |
| | | | | | | |
| :heavy_check_mark: | 395ebbfa | [#25](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/25/) | 2020-02-21 12:19:18 | ~33 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200221-121844-395ebb.zip) |
| :heavy_check_mark: | 395ebbfa | [#25](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/25/) | 2020-02-21 12:22:03 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200221-121944-395ebb.zip) |
| :heavy_check_mark: | 395ebbfa | [#25](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/25/) | 2020-02-21 12:25:02 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200221-121840-395ebb.aar) |
| | | | | | | |
| :heavy_check_mark: | cb9a66b4 | [#26](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/26/) | 2020-02-21 12:22:58 | ~33 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200221-122225-cb9a66.zip) |
| :heavy_check_mark: | cb9a66b4 | [#26](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/26/) | 2020-02-21 12:24:50 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200221-122223-cb9a66.zip) |
| :heavy_check_mark: | cb9a66b4 | [#26](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/26/) | 2020-02-21 12:31:23 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200221-122505-cb9a66.aar) |
| | | | | | | |
| :heavy_check_mark: | 03c09b41 | [#27](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/27/) | 2020-02-21 12:30:57 | ~33 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200221-123024-03c09b.zip) |
| :heavy_check_mark: | 03c09b41 | [#27](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/27/) | 2020-02-21 12:32:37 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200221-123022-03c09b.zip) |
| :heavy_check_mark: | 03c09b41 | [#27](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/27/) | 2020-02-21 12:37:36 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200221-123125-03c09b.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | cb69e3c1 | [#28](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/28/) | 2020-02-21 13:20:09 | ~36 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200221-131932-cb69e3.zip) |
| :heavy_check_mark: | cb69e3c1 | [#28](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/28/) | 2020-02-21 13:21:51 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200221-131930-cb69e3.zip) |
| :heavy_check_mark: | cb69e3c1 | [#28](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/28/) | 2020-02-21 13:25:38 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200221-131929-cb69e3.aar) |
| | | | | | | |
| :heavy_check_mark: | 3d0fc223 | [#29](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1861/29/) | 2020-02-21 14:25:09 | ~38 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200221-142430-3d0fc2.zip) |
| :heavy_check_mark: | 3d0fc223 | [#29](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1861/29/) | 2020-02-21 14:26:55 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200221-142428-3d0fc2.zip) |
| :heavy_check_mark: | 3d0fc223 | [#29](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1861/29/) | 2020-02-21 14:31:00 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200221-142427-3d0fc2.aar) |

oskarth commented on waku/peer.go:
```diff
@@ -277,6 +313,21 @@ func (p *Peer) bloomMatch(env *Envelope) bool {
 }
 
+func (p *Peer) topicInterestMatch(env *Envelope) bool {
+	p.topicInterestMu.Lock()
```
What does this and the other similar locks protect against? Perhaps a comment here would be useful, especially in case there are other instances where the class of race conditions (or whatever) can occur

cammellos replied:
Basically `map`s (or slices) are not thread safe, so we need to acquire a lock before accessing the datastructure, as they might get accessed them from different threads, I will leave a comment

oskarth commented on waku/waku.go:
```diff
@@ -56,15 +55,18 @@ type Bridge interface {
+	BloomFilter              []byte             // Bloom filter for topics of interest for this node
+	BloomFilterTolerance     []byte             // Bloom filter tolerated by the waku node for a limited time
+	TopicInterest            map[TopicType]bool // Topic interest for this node
+	TopicInterestTolerance   map[TopicType]bool // Topic interest tolerated by the waku node for a limited time
```
*see below*

oskarth commented on waku/waku.go:
```diff
@@ -56,15 +55,18 @@ type Bridge interface {
+	BloomFilterMode          bool               // Whether we should match against bloom-filter only
+	LightClient              bool               // Light client mode enabled does not forward messages
+	RestrictLightClientsConn bool               // Restrict connection between two light clients
+	SyncAllowance            int                // Maximum time in seconds allowed to process the waku-related messages
```
Update: looked at where it is used, makes sense. Do you think should perhaps make its way to the spec? If nothing else as a recommendation.

cammellos replied:
yes, makes sense, although I take is up to each client to decide how strict they want to be

kdeme replied:
I wouldn't specify specific timeouts, it really depends on clients and there use cases. But I guess recommendations do no harm. We currently have something like this mentioned for timeout on the handshake

oskarth commented:
Nice! Didn't go through Go code in great detail but looks reasonable, left some questions

kdeme commented on waku/doc.go:
```diff
@@ -50,13 +48,14 @@ const (
+	aesNonceLength   = 12                     // in bytes; for more info please see cipher.gcmStandardNonceSize & aesgcm.NonceSize()
+	keyIDSize        = 32                     // in bytes
+	BloomFilterSize  = 64                     // in bytes
+	MaxTopicInterest = 10000
```
In the specification this is currently set to `1000`

cammellos replied:
I think that's a bit tight for our use case, I can see how some busy clients might be listening on 500 topics, I thought we settled for 10k initially to be on the safe side

kdeme replied:
ok, you know the usage of topics better than me, but lets not forget to update this in the spec then.

kdeme commented on waku/handshake.go:
```diff
@@ -18,14 +18,61 @@ import (
+
+func (s statusOptions) WithDefaults() statusOptions {
+	if s.PoWRequirement == nil {
+		var pow uint64 = 0
```
How about we set the default of the peer its PoW to the same value as the node its default? 
Or some other default value, but not 0?

cammellos replied:
I would probably set it to a default rather than the host PoW, just to make it a bit simpler, we could use the current minpow used by the nodes? (can't remember exactly what it is, 0.0002 or something?)

adambabik replied:
What we use is practically 0. Especially in terms of calculating it. We run some tests once and I think anything below 0.2 or 0.02 was close to 0. We use [`0.002`](https://github.com/status-im/status-go/blob/44aa31398107f2dcb224cf6e689896aed0079227/protocol/transport/message.go#L9).

kdeme replied:
> I would probably set it to a default rather than the host PoW, just to make it a bit simpler, we could use the current minpow used by the nodes? (can't remember exactly what it is, 0.0002 or something?)

That's what I meant.

kdeme commented on waku/handshake.go:
```diff
@@ -18,14 +18,61 @@ import (
+	}
+
+	if s.RateLimits == nil {
+		s.RateLimits = &RateLimits{}
```
What does this default result in?

cammellos replied:
I believe it means no rate limits, but @adambabik can confirm

adambabik replied:
Correct. I will start working on full rate limiting support and the idea is to verify it using `RateLimits.IsZero() == false`.

kdeme commented on waku/peer.go:
```diff
@@ -224,7 +258,7 @@ func (p *Peer) broadcast() error {
 	bundle := make([]*Envelope, 0, len(envelopes))
 	for _, envelope := range envelopes {
-		if !p.marked(envelope) && envelope.PoW() >= p.powRequirement && p.bloomMatch(envelope) {
+		if !p.marked(envelope) && envelope.PoW() >= p.powRequirement && p.topicOrBloomMatch(envelope) {
```
Wait, do I understand correctly that it will apply both topic and bloom filter matching? It should only apply one or the other. Topics prioritised if they are set. Similar as I think is done in `topicInterestOrBloomMatch`.

cammellos replied:
I though it was doing that, as we don't set `fullNode` to true if `topicInterest` is set, but just looking at the function `BloomFilterMatch` is matches if bloom filter is nil, regardless on whether `fullNode` is set. I'll add a test to ensure the behavior is correct.

adambabik commented on waku/config.go:
```diff
@@ -22,6 +22,7 @@ package waku
 type Config struct {
 	MaxMessageSize           uint32  `toml:",omitempty"`
 	MinimumAcceptedPoW       float64 `toml:",omitempty"`
+	BloomFilterMode          bool    `toml:",omitempty"` // when true, we only match against topic interest
```
```suggestion
	BloomFilterMode          bool    `toml:",omitempty"` // when true, we match evenlopes against a bloom filter
```

adambabik commented on waku/handshake.go:
```diff
@@ -18,14 +18,61 @@ import (
 	TopicInterest        []TopicType `rlp:"key=5"`
 }
 
+func (s statusOptions) WithoutDefaults() statusOptions {
```
This is a bit odd. Can't we create `statusOptions` without defaults in the first place?

cammellos replied:
we can, it's useful for the testing, as it compares the size and so we need to strip out all the `zero` options, some of them are dynamically set, and it's a lot of `ifs` :) , in the actual code we use it only once, I would still encapsulate that logic in a function though, but we can create it without the defaults

adambabik commented on waku/handshake.go:
```diff
@@ -18,14 +18,61 @@ import (
+	return s
+}
+
+func (s statusOptions) WithDefaults() statusOptions {
```
I think this is a bit confusing because a default state for this node is different from a default state for a peer. Maybe instead we could have two functions.

adambabik commented on go.mod:
```diff
@@ -42,15 +43,16 @@ require (
 	github.com/status-im/migrate/v4 v4.6.2-status.2
 	github.com/status-im/rendezvous v1.3.0
 	github.com/status-im/status-go/extkeys v1.1.0
-	github.com/status-im/status-go/waku v1.2.0
```
Where did it go?

adambabik commented on waku/handshake.go:
```diff
@@ -48,24 +95,34 @@ var keyFieldIdx = func() map[string]int {
-		if !ok {
-			continue
+		field := v.Field(i)
+		if !field.IsNil() {
```
```suggestion
		if field.CanInterface() {
```
Might be even safer as it will never panic.

cammellos replied:
I think it looks like it does something slightly different, did you mean something like:
`if field.CanInterface() && !field.IsNil()` ? i.e is the field public and is not nil?

adambabik replied:
Sorry, I misunderstood `CanInterface`. We only have exported fields here so we can do both but `!field.IsNil()` is good as well 👍

adambabik commented on waku/peer.go:
```diff
@@ -94,17 +100,21 @@ func (p *Peer) handshake() error {
-		errc <- p2p.SendItems(p.ws, statusCode, ProtocolVersion, opts)
+		pow := p.host.MinPow()
+		if pow != 0 {
+			opts.SetPoWRequirementFromF(pow)
```
Shouldn't we examine other options as well? For example, if we use bloom mode, we shouldn't even set `topic-interest`, right?

cammellos replied:
currently that's not an issue, as `topicInterest`and bloom are never used in conjunction (setting one sets the other to `nil`), but we can be explicit about it

adambabik replied:
Right but when I look at this I would wonder: "I use bloom mode but in the code I still propagate `topic-interest`. Why is it so? Shouldn't I only send my bloom filter" :) 

Also, when receiving options we do `peerOptions.TopicInterest != nil`. But if we always send `topic-interest`, after RLP decoding it won't be `nil`, it will be an empty slice so we never reach `else if peerOptions.BloomFilter != nil `. Did I miss anything?

adambabik commented on waku/peer.go:
```diff
@@ -277,6 +315,29 @@ func (p *Peer) bloomMatch(env *Envelope) bool {
+
+	if topicInterestMode {
+		return p.topicInterestMatch(env)
+	} else {
```
This `else` can be omitted.

adambabik commented on waku/peer.go:
```diff
@@ -277,6 +315,29 @@ func (p *Peer) bloomMatch(env *Envelope) bool {
+	return p.fullNode || p.topicInterest[env.Topic]
+}
+
+func (p *Peer) topicOrBloomMatch(env *Envelope) bool {
```
I would explain here or somewhere else how we treat `len(topicInterest) == 0` and zero bloom filter. It is not obvious looking at the code.

adambabik commented on waku/waku.go:
```diff
@@ -1203,6 +1275,62 @@ func (w *Waku) add(envelope *Envelope, isP2P bool) (bool, error) {
+	w.settingsMu.RUnlock()
+
+	if topicInterestMode {
+		topicInterestMatch, err := w.topicInterestMatch(envelope)
```
```suggestion
		match, err := w.topicInterestMatch(envelope)
```

adambabik commented on waku/waku.go:
```diff
@@ -1203,6 +1275,62 @@ func (w *Waku) add(envelope *Envelope, isP2P bool) (bool, error) {
+		if err != nil {
+			return false, err
+		}
+		if topicInterestMatch {
```
```suggestion
		if match {
```

adambabik commented on waku/waku.go:
```diff
@@ -815,7 +899,19 @@ func (w *Waku) GetSymKey(id string) ([]byte, error) {
 	s, err := w.filters.Install(f)
 	if err == nil {
-		w.updateBloomFilter(f)
+		w.settingsMu.RLock()
```
I would create a new method and have here only something like this:
```go
w.updateSettingsForFilter(f)
```

Is there a code where we update bloom filter/topic interest after unsubscribing from a filter?

adambabik commented on waku/waku.go:
```diff
@@ -256,6 +261,71 @@ func (w *Waku) SetBloomFilter(bloom []byte) error {
+	w.notifyPeersAboutTopicInterestChange(topicInterest)
+
+	go func() {
+		// allow some time before all the peers have processed the notification
```
This should exit early if Waku is stopped. There is a channel to handle it.

adambabik commented on waku/waku.go:
```diff
@@ -256,6 +261,71 @@ func (w *Waku) SetBloomFilter(bloom []byte) error {
+// updateTopicInterest adds a new topic interest
+// and informs the peers
+func (w *Waku) updateTopicInterest(f *Filter) error {
+
```
```suggestion
```

adambabik reviewed (commented):
One more general thing. We distinguish explicitly `fullNode`. Should we do the same with `emptyNode`? Currently, it's a bit hard to navigate how the logic behaves when we have an `emptyNode`.

cammellos commented:
> One more general thing. We distinguish explicitly fullNode. Should we do the same with emptyNode
I can try to see if it makes it better, I would not introduce another piece of state as that requires further state management, although checking for `emptyNode` is a bit redundant for example (just for topic-interest for simplicity):

`emptyNode() { return !m.fullNode && len(m.topicInterest) == 0}`
`matchTopic(topic) { return m.fullNode || !m.emptyNode() && m.topicInterest[topic]}`

So the check for `emptyNode` in this case is redundant, although more explicit, what do you think?

cammellos committed:
Add status-option code

This commits changes the behavior of waku introducing a new status-code,
`2`, that replaces the current single options codes.

adambabik commented on waku/waku.go:
```diff
@@ -1410,6 +1556,31 @@ func (w *Waku) expire() {
 	}
 }
 
+func (w *Waku) ToStatusOptions() statusOptions {
```
Do we need to export it?

adambabik commented:
@cammellos re the empty node. I think it is fine if we go implicit but somewhere in the code (close to `fullNode` prop or in the setters) we should add a comment explaining how an empty node situation is handled and what "empty node" means. Especially, saying that it is a perfectly valid scenario.

cammellos commented:
@adambabik @kdeme all the feedback should be addressed, could you please review again? thanks!

adambabik reviewed (approved):
LGTM. Please check why the CI fails. Also, let's make sure `waku/` is tested in the CI pipeline. 

`whisper/` after removal of `go.mod` might also start to be tested but as far as I remember the tests were not passing after forking it from go-ethereum. I will take a look.

cammellos committed:
linting

adambabik commented on waku/waku_test.go:
```diff
@@ -278,7 +278,6 @@ func TestSymKeyManagement(t *testing.T) {
 	var err error
 	var k1, k2 []byte
 	w := New(nil, nil)
-	id1 := string("arbitrary-string-1")
```
This is causing tests to fail. 
```
[2020-02-20T11:54:35.582Z] # github.com/status-im/status-go/waku [github.com/status-im/status-go/waku.test]
[2020-02-20T11:54:35.582Z] waku/waku_test.go:283:2: undefined: id1
[2020-02-20T11:54:35.582Z] waku/waku_test.go:288:24: undefined: id1
```

They pass on develop for waku. The tests for Whisper fail on develop (confirmed) and I will fix them.
